### PR TITLE
Tab List Modification

### LINKED ISSUE #67: Tab list modification support

### PR DESCRIPTION
**Pull request by Desetude**
This PR adds the ability to read, modify, add and remove entries from a player's tab list.
I will add the documentation once all the other parts of the PR have been approved.
Resolves #67.

### CODE REVIEW

File: api/src/main/java/com/velocitypowered/api/proxy/Player.java [hunk 1]
```diff
@@ -84,17 +85,7 @@ default void sendMessage(@NonNull Component component) {
      */
     void setGameProfileProperties(List<GameProfile.Property> properties);
     
-    /**
-     * Sets the tab list header and footer for the player.
-     * @param header the header component
-     * @param footer the footer component
-     */
-    void setHeaderAndFooter(Component header, Component footer);
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/proxy/Player.java]: I would suggest retaining these methods here to limit API breakage.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java [hunk 2]
```diff
@@ -172,30 +179,23 @@ public void sendMessage(@NonNull Component component, @NonNull MessagePosition p
     public ConnectionRequestBuilder createConnectionRequest(@NonNull RegisteredServer server) {
         return new ConnectionRequestBuilderImpl(server);
     }
-    
+
     @Override
     public List<GameProfile.Property> getGameProfileProperties() {
         return this.profile.getProperties();
     }
-
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java]: No reformatting changes please.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java [hunk 3]
```diff
@@ -172,30 +179,23 @@ public void sendMessage(@NonNull Component component, @NonNull MessagePosition p
...
         Preconditions.checkNotNull(properties);
         this.profile = new GameProfile(profile.getId(), profile.getName(), properties);
     }
 
     @Override
-    public void setHeaderAndFooter(@NonNull Component header, @NonNull Component footer) {
-        Preconditions.checkNotNull(header, "header");
-        Preconditions.checkNotNull(footer, "footer");
-        connection.write(HeaderAndFooter.create(header, footer));
+    public TabList getTabList() {
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java]: If you change this to `VelocityTabList` you can avoid a cast.

File: api/src/main/java/com/velocitypowered/api/proxy/player/TabListEntry.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    Optional<Component> getDisplayName();
+
+    TabListEntry setDisplayName(@Nullable Component displayName);
+
+    int getLatency();
+
+    TabListEntry setLatency(int latency);
+
+    int getGamemode();
```
Comment (kashike) [inline api/src/main/java/com/velocitypowered/api/proxy/player/TabListEntry.java]: `getGameMode`

File: api/src/main/java/com/velocitypowered/api/proxy/player/TabListEntry.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    TabListEntry setGamemode(int gamemode);
```
Comment (kashike) [inline api/src/main/java/com/velocitypowered/api/proxy/player/TabListEntry.java]: `setGameMode`

File: proxy/src/main/java/com/velocitypowered/proxy/protocol/StateRegistry.java [hunk 6]
```diff
@@ -118,6 +118,11 @@
                     map(0x47, MINECRAFT_1_12, true),
                     map(0x48, MINECRAFT_1_12_1, true),
                     map(0x4B, MINECRAFT_1_13, true));
+            CLIENTBOUND.register(PlayerListItem.class, PlayerListItem::new,
+                    map(0x38, MINECRAFT_1_8, false),
+                    map(0x2D, MINECRAFT_1_9, false),
+                    map(0x2E, MINECRAFT_1_12_1, false),
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/protocol/StateRegistry.java]: You forgot to specify what this packet's ID for `MINECRAFT_1_12` would be.

Comment (Desetude) [reply proxy/src/main/java/com/velocitypowered/proxy/protocol/StateRegistry.java]: Fixed.

File: proxy/src/main/java/com/velocitypowered/proxy/tablist/VelocityTabList.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+public class VelocityTabList implements TabList {
+    private final MinecraftConnection connection;
+    private final Map<UUID, TabListEntry> entries = new HashMap<>();
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/tablist/VelocityTabList.java]: `ConcurrentHashMap` should be used here - there's the possibility of concurrent access.

**Comment by astei** [comment]
Also, if you could clear the tab list on server transition, that would be awesome!

**Comment by astei** [comment]
Looks good now. Please add Javadoc.

**Comment by astei** [comment]
Thanks for the contribution! 👍

**Pull request merged**